### PR TITLE
New widget - datepicker

### PR DESCRIPTION
**Pull request by ar4s**
I added one new widget (``sc-datepicker``) and replaced old ``menucalendar``
by brand new ``sc-calendar``. Both widets are compatible with Scrooge by
``menuStatsAdapterCtrl``.

### CODE REVIEW

**Comment by mkurek** [comment]
keep in mind to include later datepicker in bower config: https://github.com/allegro/ralph_pricing/pull/376

File: src/ralph_scrooge/media/scrooge/app/common/controller.js [hunk 1]
```diff
@@ -2,14 +2,28 @@
...
+		$scope.stats = stats;
+		$scope.getDictLength = function (dict) {
+			if (typeof(dict) == 'object') {
+				return Object.keys(dict).length;
+			}
+		};
+	}])
+	.controller('menuStatsAdapterCtrl', ['$scope', 'stats', function($scope, stats){
+		$scope.selected = {};
+		$scope.startDate = new Date(2014, 9, 1);
```
Comment (mkurek) [inline src/ralph_scrooge/media/scrooge/app/common/controller.js]: shouldn't this be configurable?

Comment (ar4s) [inline src/ralph_scrooge/media/scrooge/app/common/controller.js]: Yep, my mistake :)

**Comment by mkurek** [comment]
All 'costs per device' related PR should be merged to separated branch (not develop) until it's fully done - what you think?

File: src/ralph_scrooge/media/scrooge/app/allocationclient/controller.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 var scrooge = angular.module('scrooge.controller.allocationclient', []);
 
-scrooge.controller('allocationClientCtrl', ['$scope', '$routeParams', '$http', 'menuService', 'menuCalendar', 'stats', function ($scope, $routeParams, $http, menuService, menuCalendar, stats) {
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/allocationclient/controller.js]: I know it's js, and we don't use pep8, but isn't it a little too much?

File: src/ralph_scrooge/media/scrooge/app/app.js [hunk 3]
```diff
@@ -37,19 +37,19 @@ app.config(['$routeProvider', '$httpProvider',
         });
         $routeProvider
             .when('/components/', {
-                templateUrl: '/static/scrooge/partials/components.html',
+                templateUrl: STATIC_URL + 'scrooge/partials/components.html',
                 controller: 'componentsCtrl',
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/app.js]: Trailing commas can cause problems in JS

File: src/ralph_scrooge/media/scrooge/app/app.js [hunk 4]
```diff
@@ -37,19 +37,19 @@ app.config(['$routeProvider', '$httpProvider',
...
             })
             .when('/allocation/client/', {
-                templateUrl: '/static/scrooge/partials/allocationclient.html',
+                templateUrl: STATIC_URL + 'scrooge/partials/allocationclient.html',
                 controller: 'allocationClientCtrl',
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/app.js]: ditto

File: src/ralph_scrooge/media/scrooge/app/app.js [hunk 5]
```diff
@@ -37,19 +37,19 @@ app.config(['$routeProvider', '$httpProvider',
...
             })
             .when('/allocation/admin/', {
-                templateUrl: '/static/scrooge/partials/allocationadmin.html',
+                templateUrl: STATIC_URL + 'scrooge/partials/allocationadmin.html',
                 controller: 'allocationAdminCtrl',
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/app.js]: ditto

File: src/ralph_scrooge/media/scrooge/app/app.js [hunk 6]
```diff
@@ -37,19 +37,19 @@ app.config(['$routeProvider', '$httpProvider',
...
             })
             .when('/costcard/', {
-                templateUrl: '/static/scrooge/partials/costcard.html',
+                templateUrl: STATIC_URL + 'scrooge/partials/costcard.html',
                 controller: 'costCardCtrl',
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/app.js]: ditto

File: src/ralph_scrooge/media/scrooge/app/common/controller.js [hunk 7]
```diff
@@ -2,14 +2,37 @@
...
-    $scope.menuCalendar = menuCalendar;
-    $scope.stats = stats;
-    $scope.getDictLength = function (dict) {
-        if (typeof(dict) == 'object') {
-            return Object.keys(dict).length;
-        }
-    };
-}]);
+scrooge
+	.controller('mainCtrl', ['$scope', '$routeParams', 'menuService', 'stats', function ($scope, $routeParams, menuService, stats) {
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/common/controller.js]: Break it

File: src/ralph_scrooge/media/scrooge/app/common/directive.js [hunk 8]
```diff
@@ -2,10 +2,200 @@
...
+scrooge
+	.directive('tabsdirective', function() {
+	    return {
+	        restrict: 'EACM',
+	        templateUrl: '/static/scrooge/partials/tabs.html',
+	        replace: true,
+	    };
+	})
+	.directive('scCalendar', ['STATIC_URL', function(STATIC_URL) {
+		var isLeapYear = function (year) {
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/common/directive.js]: Do we have to implement shit like this? Can't we use e.g. moment.js

File: src/ralph_scrooge/media/scrooge/app/component/controller.js [hunk 9]
```diff
@@ -2,7 +2,7 @@
 
 var scrooge = angular.module('scrooge.controller.component', []);
 
-scrooge.controller('componentsCtrl', ['$scope', '$routeParams', 'menuService', 'menuCalendar', 'stats',  function ($scope, $routeParams, menuService, menuCalendar, stats) {
+scrooge.controller('componentsCtrl', ['$scope', '$routeParams', 'menuService', 'stats',  function ($scope, $routeParams, menuService, stats) {
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/component/controller.js]: Break

File: src/ralph_scrooge/media/scrooge/app/costcard/controller.js [hunk 10]
```diff
@@ -2,8 +2,7 @@
 
 var scrooge = angular.module('scrooge.controller.costcard', []);
 
-scrooge.controller('costCardCtrl', ['$scope', '$routeParams', 'menuService', 'menuCalendar', 'stats',  function ($scope, $routeParams, menuService, menuCalendar, stats) {
-    // Base configuration
+scrooge.controller('costCardCtrl', ['$scope', '$routeParams', 'menuService', 'stats',  function ($scope, $routeParams, menuService, stats) {    // Base configuration
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/app/costcard/controller.js]: ditto

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 11]
```diff
@@ -0,0 +1,1681 @@
...
+		var today = new Date();
+		return UTCDate(today.getFullYear(), today.getMonth(), today.getDate());
+	}
+	function alias(method){
+		return function(){
+			return this[method].apply(this, arguments);
+		};
+	}
+
+	var DateArray = (function(){
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: What does it do that an ordinary array cannot?

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 12]
```diff
@@ -0,0 +1,1681 @@
...
+		this.dates = new DateArray();
+		this.viewDate = UTCToday();
+		this.focusDate = null;
+
+		this._process_options(options);
+
+		this.element = $(element);
+		this.isInline = false;
+		this.isInput = this.element.is('input');
+		this.component = this.element.is('.date') ? this.element.find('.add-on, .input-group-addon, .btn') : false;
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: Break

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 13]
```diff
@@ -0,0 +1,1681 @@
...
+			o.weekStart %= 7;
+			o.weekEnd = ((o.weekStart + 6) % 7);
+
+			var format = DPGlobal.parseFormat(o.format);
+			if (o.startDate !== -Infinity){
+				if (!!o.startDate){
+					if (o.startDate instanceof Date)
+						o.startDate = this._local_to_utc(this._zero_time(o.startDate));
+					else
+						o.startDate = DPGlobal.parseDate(o.startDate, format, o.language);
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: Long as fsck

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 14]
```diff
@@ -0,0 +1,1681 @@
...
+				el.off(ev, ch);
+			}
+		},
+		_buildEvents: function(){
+			if (this.isInput){ // single input
+				this._events = [
+					[this.element, {
+						focus: $.proxy(this.show, this),
+						keyup: $.proxy(function(e){
+							if ($.inArray(e.keyCode, [27,37,39,38,40,32,13,9]) === -1)
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: ditto

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 15]
```diff
@@ -0,0 +1,1681 @@
...
+					}]
+				];
+			}
+			else if (this.component && this.hasInput){ // component: input + button
+				this._events = [
+					// For components that are not readonly, allow keyboard nav
+					[this.element.find('input'), {
+						focus: $.proxy(this.show, this),
+						keyup: $.proxy(function(e){
+							if ($.inArray(e.keyCode, [27,37,39,38,40,32,13,9]) === -1)
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: ditto

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 16]
```diff
@@ -0,0 +1,1681 @@
...
+			this._detachEvents();
+			this._detachSecondaryEvents();
+			this.picker.remove();
+			delete this.element.data().datepicker;
+			if (!this.isInput){
+				delete this.element.data().date;
+			}
+		},
+
+		_utc_to_local: function(utc){
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: Reinventing the wheel again. http://momentjs.com/timezone/docs/

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 17]
```diff
@@ -0,0 +1,1681 @@
...
+				calendarHeight = this.picker.outerHeight(),
+				visualPadding = 10,
+				windowWidth = $window.width(),
+				windowHeight = $window.height(),
+				scrollTop = $window.scrollTop();
+
+			var parentsZindex = [];
+			this.element.parents().each(function() {
+				var itemZIndex = $(this).css('z-index');
+				if ( itemZIndex !== 'auto' && itemZIndex !== 0 ) parentsZindex.push( parseInt( itemZIndex ) );
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: Lotsa lotsa too long lines

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 18]
```diff
@@ -0,0 +1,1681 @@
...
+			}
+
+			// auto y orientation is best-situation: top or bottom, no fudging,
+			// decision based on which shows more of the calendar
+			var yorient = this.o.orientation.y,
+				top_overflow, bottom_overflow;
+			if (yorient === 'auto'){
+				top_overflow = -scrollTop + offset.top - calendarHeight;
+				bottom_overflow = scrollTop + windowHeight - (offset.top + height + calendarHeight);
+				if (Math.max(top_overflow, bottom_overflow) === bottom_overflow)
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: yorient = (Math.max(top_overflow, bottom_overflow) === bottom_overflow) ? 'top' : 'bottom'

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 19]
```diff
@@ -0,0 +1,1681 @@
...
+
+		click: function(e){
+			e.preventDefault();
+			var target = $(e.target).closest('span, td, th'),
+				year, month, day;
+			if (target.length === 1){
+				switch (target[0].nodeName.toLowerCase()){
+					case 'th':
+						switch (target[0].className){
+							case 'datepicker-switch':
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: Nesting shit just got real

File: src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js [hunk 20]
```diff
@@ -0,0 +1,1681 @@
...
+				clsName: 'months',
+				navFnc: 'FullYear',
+				navStep: 1
+			},
+			{
+				clsName: 'years',
+				navFnc: 'FullYear',
+				navStep: 10
+		}],
+		isLeapYear: function(year){
```
Comment (zefciu) [inline src/ralph_scrooge/media/scrooge/js/bootstrap-datepicker.js]: Seen this code somewhere before...

**Pull request closed without merging**